client/db/types.go

```diff
@@ -212,10 +212,51 @@ func (m *MetaMatch) ID() []byte {
 	return hashKey(append(m.Match.MatchID[:], m.Match.OrderID[:]...))
 }
 
+// Revoke marks this match as revoked and retires the match if no further
+// action is required. Recovery actions may be required for a revoked match
+// if the client has broadcasted a swap contract and has not redeemed the
+// counter-party's contract or refunded own swap.
+// Returns whether or not the match is retired by this revocation action.
+func (m *MetaMatch) Revoke(fromServer bool) bool {
+	m.MetaData.Proof.serverRevoked = fromServer
+	m.MetaData.Proof.selfRevoked = !fromServer
```

Comment:
I think this is problematic as many of such matches will start out as server revoked and get switched by the refund or auto-redeem functions to self-revoked.

Reply:
I wonder why that should happen though. If it was already server revoked, no need to self-revoke. Although if a server revoke comes after a self-revoke, it'd be necessary to update that (e.g. contract locktime elapses and client auto-refunds before server revokes the match).

In any case, it'll be confusing to have both server- and self-revoke set to true for a match.

Reply:
The server would almost certainly revoke the match before client tries a refund, then the refund code in this PR sets self-revoked, unless I'm reading it wrong.  https://github.com/decred/dcrdex/pull/957/files#diff-8962b4c526755f0f98ed9f46ae05827e8edbc43ed5671ea4a2624bdf1617496cR1854

Reply:
It does and I'm thinking, if the match is already server revoked, self-revocation should be prevented. This `Revoke` method might be the best place to prevent that.

If the match was already self-revoked, then perhaps we can allow _also_ setting server-revoked but not the reverse.

Or perhaps, could retain current behaviour of setting a second revoke if there was already an initial revoke.

Reply:
There is no required relation between `ServerRevoked` and `SelfRevoked`. We should set them as needed, regardless of what the other field is set to.